src/java/org/apache/cassandra/index/sai/utils/IndexFileUtils.java

```diff
@@ -109,11 +112,49 @@ public interface ChecksumWriter
 
     class IncrementalChecksumSequentialWriter extends SequentialWriter implements ChecksumWriter
     {
-        private final CRC32 checksum = new CRC32();
-
-        IncrementalChecksumSequentialWriter(File file)
+        /** Protects the checksum so only one Writer can update it */
+        private Guard<FileChecksum> checksumGuard;
+        /** Current (running) checksum from the beginning of the file till the current position */
+        private FileChecksum checksum;
+        /** Remembers the checksum after closing this writer */
+        private long finalChecksum;
+
+        IncrementalChecksumSequentialWriter(File file, boolean append) throws IOException
         {
             super(file, writerOption);
+
+            while (checksum == null)
+            {
+                checksumGuard = checksumCache.get(file.path(), s -> new Guard<>(new FileChecksum()));
+                checksum = checksumGuard.tryLock();
+
+                if (checksum == null)
+                    // If we're here this means some other Writer did not unlock the checksum object,
+                    // so we can't use the same checksum safely, as there is a slight probability it
+                    // is in active use. This is not necessarily a bug - e.g. it is also possible
+                    // the other writer was interrupted and the client code simply forgot to close() it.
+                    // Therefore, we'll get a new one, just to be safe.
+                    checksumCache.invalidate(file.path());
+            }
+
+            if (append)
+            {
+                var fileLength = file.length();
+                skipBytes(fileLength);
+
+                // It is possible we didn't get a good checksum.
+                // We could have gotten a zero checksum because the cache has a limited size,
+                // or the file could have been changed in the meantime by another process, in which case the
+                // footer checksum would not match.
+                // However, we don't recalculate the checksum always, because it is very costly:
+                if (checksum.fileLength != fileLength || checksum.footerChecksum != calculateFooterChecksum())
+                    recalculateFileChecksum();
```

Comment:
would be nice to have some logs here in case it doesn't work as expected either because of mismatch file length or mismatch footer checksum

Reply:
+1